soc/mediatek/mtk_adsp/soc.yml

```diff
@@ -1,6 +1,10 @@
 family:
   - name: mtk_adsp
     series:
-      - name: mtk_adsp
+      - name: mt8195_adsp
         socs:
           - name: mt8195_adsp
+      - name: mt818x_adsp
+        socs:
+          - name: mt8186_adsp
+          - name: mt8188_adsp
```

Comment:
is the actual soc `mt8186` or is it `mt8186_adsp`? If the soc is `mt8186` then adsp should be a cpu cluster, not part of the soc

Reply:
These are bigcore chips with lots of subdevices intended for consumer Linux applications.  They'll never be able to share a DTS tree with the rest of the device in any practical sense.  Where we've dealt with similar DSPs in the past (c.f. soc/Intel/inetl_adsp) there's never been an attempt to shoehorn the extra hardware in and the "SOC" has always just represented the audio hardware.

I guess I'd prefer to keep that convention for simplicity and commonality, but if you really want the change can you clarify exactly how you think it should work?  I'm grepping around and not finding an obvious example of a CPU cluster to emulate.

Reply:
Ah... Yeah, I'm going to push back pretty hard on that.  Not only is it extra typing, it actually breaks the real feature that SOC_SERIES provides.   The 8188 and 8186 DSPs are actually the same Xtensa core and share almost all the configuration bits, but they live in rather different chips with different MMIO addresses.  Forcing them to be CPU clusters breaks this and I'd need to double the various #ifdef checks, etc...

And again, it's not like doing this "right" per the Zephyr hwmv2 specification is going to matter to anyone else using the chips, because they're big Linux application processors working in their own world with their own tooling and will never care about what we're doing here in Zephyr.  Really this is very much a hobgoblin-of-little-minds situation, IMHO.  I get that there's a space for complicated modeling of hardware environments, but we absolutely shouldn't be forcing it onto all users.

Reply:
the 8188 and 8186 would be separate SOCs, the part that would become a CPU cluster is the adsp, and no on the contrary it would be the same amount typing, if you have it as you do now, you have to do `west build -b <board>/mt8186_adsp` then this would instead be `west build -b <board>/mt8186/adsp` or if the board just has a single SOC and single CPU cluster it would actually be `west build -b <board>` without the SOC/CPU cluster on the end

Reply:
> or if the board just has a single SOC and single CPU cluster it would actually be west build -b <board> without the SOC/CPU cluster on the end.

actually the CPU cluster always needs to be defined, even if there is just a single.
Perhaps we should allow users to omit the CPU cluster if there is just a single cluster, but at the time we did HWMv2, then all SoC defining CPU clusters were all having at least two.

So the correct syntax would be: `west build -b <board>/mt8186/adsp` or `west build -b <board>//adsp`.

But based on discussion in this thread, allowing users to omit the CPU clusters when there is only a single makes sense.

Reply:
from what I can see in this PR then i'm aligned with @nordicjm on this one.

Having the `adsp` modeled as a CPU cluster actually makes this clearer that the adsp is part of something bigger, and where we are not supporting this bigger, because we are not modeling the other CPU clusters.

@nordicjm is not asking you to describe or implement the extra Zephyr unsupported CPU cluster.
And as it is also pointed out, then for the end-user, all which changes in having to use a `/` instead of `_` in `<board>/mt8186/adsp`.

 > Not only is it extra typing, it actually breaks the real feature that SOC_SERIES provides. The 8188 and 8186 DSPs are actually the same Xtensa core and share almost all the configuration bits, but they live in rather different chips with different MMIO addresses. Forcing them to be CPU clusters breaks this and I'd need to double the various #ifdef checks, etc...

I'm curious here ?
Can you explain exactly what extra typing will be needed and for who ?
(Only place I see is the req to always specify CPU cluster, as mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/81154/commits/ef0e92784ff401778fcd303ff0c0fdb59f58e577#r1839968759, but I'd be hapy to address that.
As well as why you cannot share the common configuration bits ?

Reply:
OK, I give up.  Not worth a fight over what ultimately is just extra indirection in stuff that doesn't impact the build.

But to be clear: MT8186 and MT8188 are not really "family-related" devices.  I mean, no more so than they are to the other members of the series.  Different AP core counts, different market targeting, etc...   But they **happen** to share exactly the same Audio DSP design (modulo, again, stuff like bus addresses).  So I can very reasonably pretend that they share the same SOC series when looking only at audio.  That gets pretty strained when the config file starts to pretend it represents the whole chip (which it doesn't of course, and never will).

Now, sure, I can come up with my own abstraction for this relationship that isn't the one baked into hwmv2.  It's just, as I said, more typing.  For IMHO essentially zero value.

Reply:
interesting discussion and confusing as well.  I thought the cpucluster would be an ARM things and we would not have to deal with it anywhere.  Having read the documentation for this multiple time does not help much https://docs.zephyrproject.org/latest/glossary.html#term-CPU-cluster in the context of adsp:

![image](https://github.com/user-attachments/assets/8f19205b-9c8f-4265-be70-df9dfb9cc3bb)

Reply:
Yeah.  In this case the distinction only makes sense when viewing the whole chip.  So you might have (to choose a less obscure device) an "intel_arrow_lake" SOC, with a cluster of "lion_cove" application processors, a cluster of "intel_ace15" audio DSPs, and in theory stuff like the camera processor and PMC and CSME and anything else that could run Zephyr.

Except that in practice all anyone cares about with this submission is the single Audio DSP core,  as the AP's on these chips are running Linux and live in an entirely different tooling world.  Which is basically the way "ancient Zephyr" configured its devices, with no need to track board-vs-soc-vs-... But alas the "simple" design is no longer simple.

Reply:
yes, exactly. And bringing in SMP into the terminology does not help much. Based on the above someone will now ask us to change all the intel IP definitions we have as part of the SoC product as cpu clusters, the shoehorning the cpu cluster terminology on such usages is an overkill IMO.

Reply:
I mean the point of hwmv2 was to have all these properties defined properly, it's not only an ARM thing. If there is only one core in a device then it doesn't make sense to add the CPU cluster but if there's more than one core then it does, you originally added the original hwmv1 board with `_adsp` when you didn't have to. And as said, if you want a common Kconfig (which can be named anything really so long as it's not using a reserved hwmv2 name like `SOC_SERIES_*`) which 2 or more SoCs select for usage in Kconfig files or drivers to allow sharing of common drivers then you can do that. They don't need to be part of the same soc series even, it's just Kconfig after all

Reply:
FWIW, I think this gets to a conflicting definition for "proper".  You want all the boxes checked to fit the config into a scheme that's designed to match every possible device.  I just want to build audio firmware for a handful of devices, and know for a fact that no one is ever going to care about Zephyr configs for the big Cortex-A/X cores that happen to share space on the silicon layout.  Making me jump through hoops and understand a complicated scheme seems... "improper".

It's the Alan Kay quote, basically.  The ideal for an API design is "Simple things should be simple, complex things should be possible.".  hwmv2 makes everything complex, simple things stop being simple.

Reply:
Is it really just a cpu inside an SoC? In many cases these things are SoCs in their own right that just *happened* to be part of a larger part, with their own versioning, isa, cpu cluster, peripheral set, memory, etc.

Reply:
hwmv2 had input from and was approved by many people, including intel. I actually find it quite jarring that as the result of having ONE underscore turn to a slash in your build commands generates this much resistance, like that is literally it, which makes it fit into the hardware scheme that everyone in zephyr is using. If no-one adds support for them, no-one adds support for them, to say that no-one ever will is a bit of a crass overstatement given that there are "big" cortex-a cores supported in zephyr e.g. raspberry pi 5, seemingly people do care about running zephyr on them.